src/Analyser/MutatingScope.php

```diff
@@ -5602,8 +5603,10 @@ private function filterTypeWithMethod(Type $typeWithMethod, string $methodName):
 	{
 		if ($typeWithMethod instanceof UnionType) {
 			$newTypes = [];
+			$oneTypeHasBeenRemoved = false;
```

Comment:
I wonder whether this is a problem which should be solved with a `Type` method.

maybe we can use a `Type` method in which you pass in a callable and it returns a preserved union type (maybe such method even exists...?)

Reply:
The whole method `filterTypeWithMethod` could be introduce as something like `Type::filter` if wanted but this would "only" be a refacto. (Which may be in another PR)

There is UnionType::traverse
```
public function traverse(callable $cb): Type
	{
		$types = [];
		$changed = false;

		foreach ($this->types as $type) {
			$newType = $cb($type);
			if ($type !== $newType) {
				$changed = true;
			}
			$types[] = $newType;
		}

		if ($changed) {
			return TypeCombinator::union(...$types);
		}

		return $this;
	}
```
but it doesn't do the exact same thing ; we can still notice a `changed` check is done the same way.

Reply:
don't want to disturb you guys too much, but another idea to throw in - does `TypeTraverser::map()` help in any way here maybe? e.g. the types you don't want you could replace with a `NeverType` potentially? that should leave the outer TemplateUnionType untouched at least. that could be a replacement for the foreach, AFAIK `TypeCombinator::union()` should keep the TemplateUnionType then.

Reply:
I get it, but the variable name should be different. Something like `$unionTypeHasChanged`.

Also it's a lot of repeated code. I'd welcome some refactoring first, and then the change should be done in a single place.